src/Radicle/Storage/Types.hs

```diff
@@ -0,0 +1,130 @@
+{-# LANGUAGE StrictData #-}
+
+module Radicle.Storage.Types where
+
+import           Protolude
+
+import           Data.Aeson (FromJSON, ToJSON)
+
+import           Radicle
+
+-- | In-memory representation of the current state of a machine.
+data Machine mid idx = Machine
+    { machineId        :: mid
+    , machineLastIndex :: Maybe idx
+    , machineState     :: Bindings (PrimFns Identity)
+    , machineMode      :: ReaderOrWriter
+    }
+
+-- | Class of types which can be rendered as a valid file name.
+--
+-- On *nix, this essentially means any character except the directory separator
+-- \'/\' and NUL (\'\0\'). The 'FilePath' should also be < 251 characters.
+--
+class FilePathComponent a where
+    toFilePathComponent :: a -> FilePath
+
+-- | Constraint 'Backend'-specific machine identifiers must satisfy
+type MachineId a =
+    ( Eq       a
+    , Show     a
+    , Typeable a
+    , Hashable a
+    , ToJSON   a
+    , FromJSON a
+    , FilePathComponent a
+    )
+
+-- | Constraint 'Backend'-specific machine indices must satisfy
+type MachineIndex a = (ToJSON a, FromJSON a)
+
+-- | Indicates if the 'Machine' is writable by the local daemon.
+data ReaderOrWriter = Reader | Writer
+    deriving (Generic)
+
+instance ToJSON ReaderOrWriter
+instance FromJSON ReaderOrWriter
+
+type Inputs  = Expressions
```

Comment:
There isn't really a difference between inputs and outputs, type-wise. It's just that we would _talk_ about inputs and outputs all the time, so that reads nicer. Did I actually use `Outputs` anywhere?

Reply:
Ya, you use it in a few places: https://github.com/radicle-dev/radicle/pull/670/files#diff-34853626d24fdb9f6203c660f493ae4aR56

I'm ok with type-aliasing in this case :)